bigquery_etl/cli/backfill.py

```diff
@@ -488,6 +486,60 @@ def complete(ctx, qualified_table_name, sql_dir, project_id):
     )
 
 
+def _copy_backfill_staging_to_prod(
+    backfill_staging_table: str,
+    qualified_table_name: str,
+    client: bigquery.Client,
+    entry: Backfill,
+    table_metadata: Metadata,
+):
+    """Copy backfill staging table to prod based on table metadata and backfill config.
+
+    If table is
+       un-partitioned: copy the entire staging table to production.
+       partitioned: determine and copy each partition from staging to production.
+    """
+    partitioning_type = None
+    if table_metadata.bigquery and table_metadata.bigquery.time_partitioning:
+        partitioning_type = table_metadata.bigquery.time_partitioning.type
+
+    if partitioning_type is None:
+        _copy_table(backfill_staging_table, qualified_table_name, client)
+    else:
+        backfill_date_range = BackfillDateRange(
+            entry.start_date,
+            entry.end_date,
+            excludes=entry.excluded_dates,
+            range_type=partitioning_type,
+        )
+        # If date_partition_parameter isn't set it's assumed to be submission_date:
+        # https://github.com/mozilla/telemetry-airflow/blob/dbc2782fa23a34ae8268e7788f9621089ac71def/utils/gcp.py#L194C48-L194C48
+        partition_param, offset = "submission_date", 0
+        if table_metadata.scheduling:
+            partition_param = table_metadata.scheduling.get(
+                "date_partition_parameter", partition_param
+            )
+            offset = table_metadata.scheduling.get("date_partition_offset", offset)
+
+        for backfill_date in backfill_date_range:
+            if (
+                partition := get_backfill_partition(
+                    backfill_date,
+                    partition_param,
+                    offset,
+                    partitioning_type,
+                )
```

Comment:
There's some code smell here with:
File1: Defines `BackfillDateRange` and `get_backfill_partition` func separately
File2: Iterates over `BackfillDateRange` and calls the `get_backfill_partition` for each date

Could the iterator over `BackfillDateRange` yield both (date and partition)?

Reply:
Potentially yes, but getting the backfill partitions and getting the backfill dates are different pieces of functionality, see the other comment on the `_backfill_query` function. I considered making `BackfillDateRange` offset aware but (1) the non-offset date was still useful and (2) it's clear here which date is which, you don't have to remember whether the offset or the original date is first in the tuple